pinot-segment-local/src/main/java/org/apache/pinot/segment/local/io/util/ValueReaderComparisons.java

```diff
@@ -0,0 +1,174 @@
+import org.apache.pinot.segment.spi.memory.PinotDataBuffer;
+
+public class ValueReaderComparisons {
+
+    private ValueReaderComparisons() {
+    }
+
+    private static int mismatch(PinotDataBuffer dataBuffer, long startOffset, int length, ByteBuffer buffer) {
+        boolean littleEndian = dataBuffer.order() == ByteOrder.LITTLE_ENDIAN;
+        if (littleEndian) {
+            buffer.order(ByteOrder.LITTLE_ENDIAN);
+        }
+        int limit = Math.min(length, buffer.limit());
+        int loopBound = limit & ~0x3F;
+        int i = 0;
+        for (; i < loopBound; i += 8) {
+            long ours = dataBuffer.getLong(startOffset + i);
+            long theirs = buffer.getLong(i);
+            if (ours != theirs) {
+                long difference = ours ^ theirs;
+                return i + ((littleEndian
+                        ? Long.numberOfTrailingZeros(difference)
+                        : Long.numberOfLeadingZeros(difference)) >>> 3);
+            }
+        }
+        for (; i < limit; i++) {
+            byte ours = dataBuffer.getByte(startOffset + i);
+            byte theirs = buffer.get(i);
+            if (ours != theirs) {
+                return i;
+            }
+        }
+        return -1;
+    }
+
+    static int compareUtf8Bytes(PinotDataBuffer dataBuffer, long startOffset, int length, boolean padded,
+                           byte[] bytes) {
+        // can use MethodHandles.byteArrayViewVarHandle here after dropping JDK8
+        ByteBuffer buffer = ByteBuffer.wrap(bytes);
+        int mismatchPosition = mismatch(dataBuffer, startOffset, length, buffer);
+        if (mismatchPosition == -1) {
+            // need to figure out whether the unpadded string is longer than the parameter or not
+            if (padded && bytes.length < length) {
+                if (bytes.length == 0) {
+                    // just need nonzero first byte
+                    return dataBuffer.getByte(startOffset) == 0 ? 0 : 1;
+                } else if (bytes[bytes.length - 1] == 0) {
+                    // we can't store zero padding, but can have it in parameters, check the last byte
+                    return dataBuffer.getByte(startOffset + bytes.length - 1) == 0 ? -1 : 0;
+                } else {
+                    // check if the stored string continues beyond the length of the parameter
+                    return dataBuffer.getByte(startOffset + bytes.length) == 0 ? 0 : 1;
+                }
+            } else {
+                // then we know the length precisely or know that the parameter is at least as long as we can store
+                return length - bytes.length;
+            }
+        }
+        byte ourByte = dataBuffer.getByte(startOffset + mismatchPosition);
+        if (ourByte == 0) {
+            // then we are storing the empty string
+            return -1;
+        }
+        byte theirByte = bytes[mismatchPosition];
+        if (theirByte == 0) {
+            return 1;
+        }
```

Comment:
I don't think we need these special handling since they are not very common case. We may change `first > 0` to `first >= 0` to handle the 0 byte in `compareUtf8()`

Reply:
Yes, that's a good simplification.